### PR TITLE
[Closed] Update encumbrance display to be complete and much more correct

### LINKED ISSUE #16833: The encumbrance penalties shown in Player Info don't match the real ones
Almost all of the encumbrance penalties shown under View Player Info > Encumbrance and Warmth are either out of sync with the actual penalties or are just a sentence with no numbers in it.

In the long run all the encumbrance code should probably be changed to share functions with player::disp_info() to keep this from happening again.

### PR DESCRIPTION
**Pull request by ghost**
Most of the listed penalties in the ENCUMBRANCE AND WARMTH tab of Player info were either inaccurate, missing, or just a sentence with no numbers. They also used separate calculations for the listed penalties and the actual penalties. This PR aims to fix most of that by using the same functions to calculate both the descriptions in Player Info ENCUMBRANCE AND WARMTH and the actual penalties elsewhere in the code.

fixes #16833 

There is still a slight issue with rounding differences, because the display can only show a single body part at a time while the actual penalties can use integer math to combine multiple body parts. This is still a vast improvement over what was previously there, which also suffered from rounding differences.

Some encumbrance penalties also reflect minimum or maximum caps, for example a sufficiently good swimmer might see low penalties listed if they are near or at the swim speed cap of 30 movement points. By comparison, before this update, the listed swim speed penalties were only a small fraction of the actual totals, since they ignored the huge penalty from equipped volume.

### CODE REVIEW

**Comment by illi-kun** [comment]
It's not good to glue build fix to some other code changes. I'll merge your fix of unrelated astyle regression separately.

**Comment by Crispy6** [comment]
I couldn't even compile it locally without the fix. It's not good to be dead in the water either. I'll pull them back out, though.

**Comment by Crispy6** [comment]
Ok, took the unrelated fix back out.

**Comment by kevingranade** [comment]
Most instances of "sentences with no numbers" are intended.

File: src/player.cpp [hunk 1]
```diff
@@ -1598,6 +1597,19 @@ void player::recalc_speed_bonus()
 
 int player::run_cost(int base_cost, bool diag) const
 {
+    return move_cost_encumbrance( base_cost, diag );
```
Comment (Coolthulhu) [inline src/player.cpp]: This rename is weird and misleading: `move_cost_encumbrance` sounds like extra move cost due to encumbrance, not the entire run cost.

`move_cost_encumbrance` should be just the part at the very bottom of the function. You can split the function into `move_cost_base` and `move_cost_encumbrance` if you need to exclude body parts (with `move_cost_base` not having body part exclusion arguments).

File: src/player.cpp [hunk 2]
```diff
@@ -1732,8 +1759,15 @@ int player::run_cost(int base_cost, bool diag) const
     return int(movecost);
 }
 
+
 int player::swim_speed() const
 {
+    return swim_speed_encumbrance();
```
Comment (Coolthulhu) [inline src/player.cpp]: Same as with move cost - `swim_speed_encumbrance` shouldn't include the base cost code, only the encumbrance-related one.

**Comment by Crispy6** [comment]
I'll revert three of those sentences back to:
- "Head encumbrance has no effect; it simply limits how much you can put on. "
- "Covering your mouth will make it more difficult to breathe and catch your breath. "
- "Reduces the speed at which you can handle or manipulate items. " (listed under hands)

There is a consistency issue with the last one of those sentences. The eye encumbrance had displayed perception modifiers for firing ranged weapons and when throwing items. Both of the values displayed were completely wrong after changes to ranged weapons and throwing, but they were there. The arms have just as much effect on ranged weapon accuracy as eyes. Likewise the hands have just as much effect on throwing as eyes.

For ranged accuracy and throwing, which were the reason I started this whole update, should they be displayed like eye encumbrance or hidden like arm encumbrance?

File: src/player.cpp [hunk 3]
```diff
@@ -2342,6 +2396,268 @@ std::string dodge_skill_text(double mod)
...
+                                              both_sides) );
+    s += string_format( ngettext( "Swimming costs %+d movement point. ",
+                                  "Swimming costs %+d movement points. ",
+                                  ( swim_cost == 0 ? 2 : swim_cost ) ), swim_cost );
+}
+
+void player::exclude_body_parts( int &encumb1, const body_part part1, int &encumb2,
+                                 const body_part part2, int &encumb3,const body_part part3,
+                                 const bool exclude, const body_part excluded_part,
+                                 const bool both_sides ) const
```
Comment (Coolthulhu) [inline src/player.cpp]: That's unreadable, needs comments at the very least.

**Comment by Crispy6** [comment]
@Coolthulhu

I can fix the weird names for `move_cost_encumbrance` and `swim_speed_encumbrance`, but in both of those cases the actual encumbrance penalty is extremely difficult to calculate without just running through the whole function to look at the difference between encumbered speed and unencumbered speed. Swim speed is even more problematic, because equipped item volume has a huge effect on swim speed unless swim skill is 10 or higher.

One of the goals for this update was to have displayed encumbrance penalties and actual encumbrance penalties run the same code so that they would actually match instead of saying one thing but actually being something else.

Would you still like me to deliberately show inaccurate penalties that I know are inaccurate?

File: src/player.cpp (hunk 3 above)
Comment (Crispy6) [inline src/player.cpp]: It does have a very short comment in the header, but I'll add a proper description of the function and all its parameters to that.

**Comment by Crispy6** [comment]
@kevingranade 
I'll change the sentences with no numbers back.

One of them is inconsistent, though. Eyes show numbers for their effects on ranged combat and throwing (they're so out of date as to be completely wrong, but they're there). Arms just have a sentence that arm encumbrance affects accuracy with ranged weapons. Should these accuracy penalties be shown like for eyes or hidden like for arms and hands? Or is it really intended to show one and hide the other?

**Comment by Coolthulhu** [comment]
It's rather ugly to calculate the entire penalty twice and then drop a tiny bit of it in the second calculation.

It's an improvement over the duplication we have now, because the code is in one place, but I'm pretty sure it could be done better.

Note that the move cost penalty for encumbrance is calculated near the very bottom of the function and it is only multiplied by diagonal penalty and stamina penalty. You could have a function like
```
int move_cost_penalty( body_part exclude, bool both_parts, bool diag, float cost_multiplier )
```

With `cost_multiplier` being the stamina penalty.

**Comment by Crispy6** [comment]
For move cost I agree with you and I'll change it. Guess I got trigger happy. For swim speed, my point stands.

**Comment by Crispy6** [comment]
@kevingranade 
Reloading speed is inconsistent too. A number was displayed, which is out of date and incorrect, but the actual reloading speed was rolled into item handling in general, which was described with the sentence, "Reduces the speed at which you can handle or manipulate items".

Should the item manipulation speed be hidden or shown?

**Comment by codemime** [comment]
I strongly dislike this approach. It's an unmanageable clod of procedural mud. What you generally need is a class that would supply necessary values to the interested party. The example isn't optimised, but it reveals the proposed approach:

```c++
// An abstract class that would supply values
class values_supplier {
    public:
        virtual int encumb( body_part bp ) const = 0;
};

//  A heir for characters
class character_values_supplier : public values_supplier {
    public:
        character_values_supplier( const Character &character ) : 
            values_supplier(),
            character( character ) {};
        virtual int encumb( body_part bp ) const override {
            return character.encumb( bp );
        }
    private:
        Character &character;
};

// A heir for displaying purposes
class display_values_supplier : public character_values_supplier {
    public:
        display_values_supplier( const Character &character, body_part excluded_bp ) :
            character_values_supplier( character ),
            excluded_bp( excluded_bp ) {};

        virtual int encumb( body_part bp ) const override {
            if( bp == excluded_bp )
                return 0;
            return character_values_supplier::encumb( bp );
        }
    private:
        body_part excluded_bp;
};
```

Then you can write any formula **only once** using the abstract version of the supplier:

```c++
int stats::swim_speed( values_supplier values ) const {
   return ... * values.encumb( bp_torso ) * ...
}
```

Character use it like this:

```c++
int Character::swim_speed() {
    return stats::swim_speed( character_values_supplier( *this ) ); // Actually you need to create the object only once.
}
```

When you need to recalculate the formula, but exclude some of the body parts, you write it like this:
```c++
return stats::swim_speed( display_values_supplier( *this, bp_leg ) );
```
You can do pretty much whatever (even unit tests :). It's flexible.

**Comment by Crispy6** [comment]
@Coolthulhu 
I remember now. If calculating the move speed penalty by adding body parts instead of subtracting out one at a time, it would take a lot of really ugly code to duplicate the effect of the integer divide by 10 in
```
    movecost +=
        ( ( encumb(bp_foot_l) + encumb(bp_foot_r) ) * 2.5 +
          ( encumb(bp_leg_l) + encumb(bp_leg_r) ) * 1.5 ) / 10;
```
I’ll still break the function out of the original `player::run_cost`, but that’s basically just moving lines around from what's there now.

Architecturally I think all the encumbrance penalties should be a single number per body part with no shared rounding between them. That isn't what's there now, though and I was trying to leave the above movecost formula intact.

**Comment by Crispy6** [comment]
@codemime 
I vote against creating three new classes like that. I see no reason why the player class cannot handle everything about the encumbrance. I do note your objections, though, so I will simplify my code further.

If I had carte blanche to get rid of all shared rounding between separate body parts (which your example above seems like it would also require) I could make it better still.

**Comment by codemime** [comment]
> I vote against creating three new classes like that.

Four actually. But they will pay for themselves.

> I see no reason why the player class cannot handle everything about the encumbrance.

The main reason is that the `player` is a bloated [god object](https://en.wikipedia.org/wiki/God_object) at its best. It shouldn't be aware of UI at all. Displaying encumbrance per body part and stat diffs are certainly UI details. Neither `Character` nor `player` must know a thing about it. This function (and many others) enforces that very-very UI-specific knowledge in a very misleading (to a reader) way. It accepts **nine!** arguments and has a name of a modifier (surgical?).

```c++
        void player::exclude_body_parts( int &encumb1, const body_part part1, int &encumb2,
                                         const body_part part2, int &encumb3,const body_part part3,
                                         const bool exclude, const body_part excluded_part,
				     const bool both_sides ) const;

```
> If I had carte blanche to get rid of all shared rounding between separate body parts (which your example above seems like it would also require) I could make it better still.

I don't see why it's a problem. Could you expand on that?

**Comment by Crispy6** [comment]
Hmm, I misspoke about the encumbrance. I believe that the player class should be able to handle everything about the actual encumbrance penalty. I agree that displaying the encumbrance should be handled elsewhere, but I've already vastly increased the scope of this pull request once, I'm not going to address that here.

I'd be happy to work with you in the future on breaking up the player class into sensible classes. I'm also fine if you preferred to handle that yourself.

For the shared rounding:
`( ( encumb_foot_l + encumb_foot_r ) * 2.5 + ( encumb_leg_l + encumb_leg_r ) * 1.5 ) / 10`
is not equal to
`( encumb_foot_l * 2.5 / 10 ) + ( encumb_foot_r * 2.5 / 10 ) + ( encumb_leg_l * 2.5 / 10 ) + ( encumb_leg_r * 2.5 / 10 )`
due to rounding errors with the integer math. Even if it were all done with `double`, it would then have too _little_ rounding error compared to the original formula.

If all the math on the encumbrance penalties themselves was floating point, the two above expressions would be equal and everything would be very simple. I'm going to code that up and see how well it works.

**Comment by Crispy6** [comment]
It is coming out much cleaner using all `double` math and using arrays of coefficients for each body part, but I'm just not willing to break all the existing rounding. Even what’s in here now changed it somewhere, I think it was throwing bonus.

Between that and being told that most of the missing information is missing by design, the cost/benefit ratio for this doesn't seem good enough. I'm closing this second attempt at a PR too.

My thanks to all of you for the feedback.

**Pull request closed without merging**